### PR TITLE
[storage] Adds support for Append Blobs

### PR DESCRIPTION
**Pull request by BrianBland**
Adds PutAppendBlob and AppendBlock methods

Addresses #275

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
@BrianBland thanks for the pull request. LGTM. `extraHeaders` will let users make use of `x-ms-blob-condition-maxsize` and `x-ms-blob-condition-appendpos`, which might be frequently used in the use cases of AppendBlob.

It looks like there's something funny going on about our storage tests. @colemickens have you changed something lately?

**Comment by BrianBland** [comment]
@ahmetalpbalkan it looks like I need to change some things up to support the new signing code as of version 2015-02-21 (see https://msdn.microsoft.com/en-us/library/azure/mt584141.aspx)

**Comment by ahmetalpbalkan** [comment]
@BrianBland oh that explains the problem. We were actually planning to switch on SharedKeyLite signing (which is way simpler, doesn't have Content-Length either). If you don't mind making the change to work with `2015-02-21`, please go ahead.

**Comment by BrianBland** [comment]
Hopefully you don't mind the repeated push trial and error with the CI system. I'm temporarily locked out of my Azure account :disappointed:

**Comment by BrianBland** [comment]
@ahmetalpbalkan if you have any idea why `TestBlobSASURICorrectness` is failing with the new API version, I'm ready to fix it. I can't find any documentation suggesting why this is suddenly invalid.

**Comment by colemickens** [comment]
(Won't be near a computer for a while but I haven't changed anything for
storage tests)

**Comment by ahmetalpbalkan** [comment]
@BrianBland oh that's just comparing a precomputed signature with the existing algorithm. You just need to change the hard-coded computed value with the result of then modified algorithm.

**Comment by BrianBland** [comment]
@ahmetalpbalkan I think you're referring to `TestGetBlobSASURI`. The failing test is getting a 403 status code from the server

**Comment by BrianBland** [comment]
Apparently the canonicalizedResource for the presigned blob urls must begin with "/blob", even though [the documentation](https://msdn.microsoft.com/en-us/library/azure/dn140255.aspx) suggests that it should begin with "blob"

**Comment by ahmetalpbalkan** [comment]
@BrianBland I'll follow up on that. Might be a docs issue indeed.

File: storage/blob.go [hunk 1]
```diff
@@ -801,6 +802,49 @@ func (b BlobStorageClient) GetPageRanges(container, name string) (GetPageRangesR
...
+// PutAppendBlob initializes an empty append blob with specified name. An
+// append blob must be created using this method before appending blocks.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179451.aspx
+func (b BlobStorageClient) PutAppendBlob(container, name string, extraHeaders map[string]string) error {
+	path := fmt.Sprintf("%s/%s", container, name)
+	uri := b.client.getEndpoint(blobServiceName, path, url.Values{})
+	headers := b.client.getStandardHeaders()
+	headers["x-ms-blob-type"] = string(BlobTypeAppend)
+	headers["Content-Length"] = "0"
```
Comment (ahmetalpbalkan) [inline storage/blob.go]: @BrianBland Hmm, I don't think you will still need this header now that you updated `buildCanonicalizedString` with

```diff
+	contentLength := headers["Content-Length"]
+	if contentLength == "0" {
+		contentLength = ""
+	}
```

In fact, there should be more occurrences of `headers["Content-Length"] = "0"` in this file, I think we can remove all those if you don't mind doing this in PR as well?

Comment (BrianBland) [inline storage/blob.go]: No worries, removed all of those

**Comment by ahmetalpbalkan** [comment]
@BrianBland thanks LGTM!

**Comment by ahmetalpbalkan** [comment]
@BrianBland hi there, just circling back on the docs issue here https://msdn.microsoft.com/en-us/library/azure/dn140255.aspx I see that it says `canonicalizedresource = "/blob/...`, did you see a different value back then?

> image

**Comment by BrianBland** [comment]
I thought so, but at least it looks correct now

**Pull request merged**